examples/post-sync/main.go

```diff
@@ -0,0 +1,352 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+
+	"github.com/go-logr/logr"
+	"github.com/go-logr/zapr"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var (
+	scheme = runtime.NewScheme()
+)
+
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(v1beta1.AddToScheme(scheme))
+}
+
+// Define a new struct for the unique identifier
+type SyncID struct {
+	Name      string
+	Kind      string
+	Namespace string
+}
+
+type SyncStatusController struct {
+	client        client.Client
+	log           logr.Logger
+	statusTracker *StatusTracker
+	syncKind      string // Explicit field to indicate what kind of resource we're watching
+}
+
+func (c *SyncStatusController) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	var syncID SyncID
+	var status v1beta1.Status
+	var generation int64
+	var observedGeneration int64
+	var errs, commit string
+
+	// Use the explicit syncKind to know which type to fetch
+	switch c.syncKind {
+	case "RootSync":
+		var root v1beta1.RootSync
+		if err := c.client.Get(ctx, req.NamespacedName, &root); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+		syncID = SyncID{
+			Name:      root.Name,
+			Kind:      root.Kind,
+			Namespace: root.Namespace,
+		}
+		status = root.Status.Status
+		generation = root.Generation
+		observedGeneration = root.Status.ObservedGeneration
+		errs, commit = c.processRootSyncConditions(root.Status.Conditions)
+
+	case "RepoSync":
+		var repo v1beta1.RepoSync
+		if err := c.client.Get(ctx, req.NamespacedName, &repo); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+		syncID = SyncID{
+			Name:      repo.Name,
+			Kind:      repo.Kind,
+			Namespace: req.Namespace,
+		}
+		status = repo.Status.Status
+		generation = repo.Generation
+		observedGeneration = repo.Status.ObservedGeneration
+		errs, commit = c.processRepoSyncConditions(repo.Status.Conditions)
+
+	default:
+		c.log.Error(fmt.Errorf("unknown sync kind: %s", c.syncKind), "unrecognized syncKind")
+		return reconcile.Result{}, fmt.Errorf("unknown sync kind: %s", c.syncKind)
+	}
+
+	// Process any errors from conditions
+	if err := c.handleErrors(syncID, errs, commit, generation, observedGeneration); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Process sync status
+	if !isSuccessfulSync(status) {
+		errors, commit := extractErrorAndCommitFromStatus(status)
```

Comment:
nit: kind odd to use `errs` above and `errors` here. Looks like these are the same thing, might as well use the same variable name.

Reply:
if it's just a string, then maybe `errorMessage`

Reply:
errMessage fits more to the scenario, changed